### PR TITLE
wasm: Add support for WASI threads in CMake build system

### PR DESCRIPTION
**Pull request by kateinoigakukun**
This patch adds a new CMake option, `SWIFT_ENABLE_WASI_THREADS`, to enable building the Standard Library using WASI threads primitives (https://github.com/WebAssembly/wasi-threads). With this option, the Standard Library will be built for the new "wasm32-unknown-wasi-threads" target. We add the new triple with "-threads" environment field because the WASI threads proposal requires extra syscalls and object files compiled to "wasm32-unknown-wasi" and "wasm32-unknown-wasi-threads" are not compatible and cannot be linked together.

I chose `wasm32-unknown-wasi-threads` by following wasi-sdk: https://github.com/WebAssembly/wasi-sdk/blob/f1ebc52a74394cdf885d03bfde13899b3d5c6d2d/wasi-sdk-pthread.cmake#L9

### CODE REVIEW

**Comment by kateinoigakukun** [comment]
@swift-ci smoke test

**Comment by kateinoigakukun** [comment]
@swift-ci test WebAssembly

File: cmake/modules/SwiftConfigureSDK.cmake [hunk 1]
```diff
@@ -418,7 +422,11 @@ macro(configure_sdk_unix name architectures)
           message(FATAL_ERROR "unsupported arch for WebAssembly: ${arch}")
         endif()
         set(SWIFT_SDK_WASI_ARCH_wasm32_PATH "${SWIFT_WASI_SYSROOT_PATH}")
-        set(SWIFT_SDK_WASI_ARCH_wasm32_TRIPLE "wasm32-unknown-wasi")
+        if(SWIFT_ENABLE_WASI_THREADS)
+          set(SWIFT_SDK_WASI_ARCH_wasm32_TRIPLE "wasm32-unknown-wasi-threads")
```
Comment (MaxDesiatov) [inline cmake/modules/SwiftConfigureSDK.cmake:426]: Is there no other way to express this other than using the last triple component? Is this is the triple that Clang uses for WASI threads as well? I find it confusing that we have `wasm32-unknown-none-wasm` with the last triple component denoting the object format, and `wasm32-unknown-wasi-threads` for denoting platform feature. OTOH this is not much different from `armv7-unknown-linux-musl`?

Comment (kateinoigakukun) [reply cmake/modules/SwiftConfigureSDK.cmake:426]: I followed how wasi-sdk and wasi-libc call the target: https://github.com/WebAssembly/wasi-sdk/blob/f1ebc52a74394cdf885d03bfde13899b3d5c6d2d/wasi-sdk-pthread.cmake#L9

Target triple parser is very complicated... The fourth component of *triple* can be interpreted as environment or object file format.
https://github.com/llvm/llvm-project/blob/a872a35251b833aaddec2172710ff234236afbd8/llvm/lib/TargetParser/Triple.cpp#L1071-L1076

**Review by MaxDesiatov — APPROVED** [review]

**Review by kalebers — APPROVED** [review]

**Review by al45tair — APPROVED** [review]

**Pull request merged**